src/code/playback.c

```diff
@@ -0,0 +1,786 @@
+    s32 i;
+    AudioListItem* source;
+    AudioListItem* cur;
+    AudioListItem* dest;
+
+    for (i = 0; i < 4; i++) {
+        switch (i) {
+            case 0:
+                source = &pool->disabled;
+                dest = &gNoteFreeLists.disabled;
+                break;
+
+            case 1:
+                source = &pool->decaying;
+                dest = &gNoteFreeLists.decaying;
+                break;
+
+            case 2:
+                source = &pool->releasing;
+                dest = &gNoteFreeLists.releasing;
+                break;
+
+            case 3:
+                source = &pool->active;
+                dest = &gNoteFreeLists.active;
+                break;
+        }
+
+        for (;;) {
+            cur = source->next;
+            if (cur == source || cur == NULL) {
+                break;
+            }
+            audio_list_remove(cur);
+            audio_list_push_back(dest, cur);
+        }
+    }
+}
+
+void note_pool_fill(NotePool* pool, s32 count) {
+    s32 i;
+    s32 j;
+    Note* note;
+    AudioListItem* source;
+    AudioListItem* dest;
+
+    note_pool_clear(pool);
+
+    for (i = 0, j = 0; j < count; i++) {
+        if (i == 4) {
+            return;
+        }
+
+        switch (i) {
+            case 0:
+                source = &gNoteFreeLists.disabled;
+                dest = &pool->disabled;
+                break;
+
+            case 1:
+                source = &gNoteFreeLists.decaying;
+                dest = &pool->decaying;
+                break;
+
+            case 2:
+                source = &gNoteFreeLists.releasing;
+                dest = &pool->releasing;
+                break;
+
+            case 3:
+                source = &gNoteFreeLists.active;
+                dest = &pool->active;
+                break;
+        }
+
+        while (j < count) {
+            note = audio_list_pop_back(source);
+            if (note == NULL) {
+                break;
+            }
+            audio_list_push_back(dest, &note->listItem);
+            j++;
+        }
+    }
+}
+
+void audio_list_push_front(AudioListItem* list, AudioListItem* item) {
+    // add 'item' to the front of the list given by 'list', if it's not in any list
+    if (item->prev == NULL) {
+        item->prev = list;
+        item->next = list->next;
+        list->next->prev = item;
+        list->next = item;
+        list->u.count++;
+        item->pool = list->pool;
+    }
+}
+
+void audio_list_remove(AudioListItem* item) {
+    // remove 'item' from the list it's in, if any
+    if (item->prev != NULL) {
+        item->prev->next = item->next;
+        item->next->prev = item->prev;
+        item->prev = NULL;
+    }
+}
+
+Note* pop_node_with_value_less_equal(AudioListItem* list, s32 limit) {
+    AudioListItem* cur = list->next;
+    AudioListItem* best;
+
+    if (cur == list) {
+        return NULL;
+    }
+
+    best = cur;
+    for (; cur != list; cur = cur->next) {
```

Comment:
Could the for loop be 
```suggestion
    for (cur = list->next, best = cur; cur != list; cur = cur->next) {
```
And remove 
```
if(cur == list) {
    return NULL;
}

best = cur;
```

Reply:
I can't remove the conditional, but was able to put the rest in the for and remove best = cur;